substrate/utils/frame/remote-externalities/src/logging.rs

```diff
@@ -0,0 +1,86 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+	future::Future,
+	io::{self, IsTerminal},
+	time::Instant,
+};
+
+use spinners::{Spinner, Spinners};
+
+use super::Result;
+
+// A simple helper to time a operation with a nice spinner, start message, and end message.
+//
+// The spinner is only displayed when stdout is a terminal.
+pub(super) fn with_elapsed<F, R, EndMsg>(f: F, start_msg: &str, end_msg: EndMsg) -> Result<R>
+where
+	F: FnOnce() -> Result<R>,
+	EndMsg: FnOnce(&R) -> String,
+{
+	let timer = Instant::now();
+	let mut maybe_sp = start(start_msg);
+
+	Ok(end(f()?, timer, maybe_sp.as_mut(), end_msg))
```

Comment:
Could probably get rid of the sync variant if you wanted to and always use the async one.